src/zope/publisher/http.py

```diff
@@ -71,6 +71,37 @@ class CookieMapper(RequestDataMapper):
 class HeaderGetter(RequestDataGetter):
     _gettrname = 'getHeader'
 
+
+host_port_re = re.compile(
+    r"^(.*):([0-9]*)$", re.DOTALL)
+
+def splitport(host):
+    """Split port number off the hostname.
+
+        >>> splitport('example.com:80')
+        ('example.com', '80')
+
+        >>> splitport('localhost')
+        ('localhost', None)
+
+        >>> splitport('[::1]')
+        ('[::1]', None)
+
+        >>> splitport('[::1]:443')
+        ('[::1]', '443')
+
+        >>> splitport('localhost:')
+        ('localhost', None)
+
+    """
+    match = host_port_re.match(host)
+    if match:
+        host, port = match.groups()
+    else:
+        port = None
+    return host, port or None
```

Comment:
Interesting, some while ago I implemented the same function in Zope using the suggested `urlparse` function, see

https://github.com/zopefoundation/Zope/blob/e9f10101b8a4db64a2440e3a51561a7805ea44f5/src/ZPublisher/HTTPRequest.py#L110-L128

Maybe we could use your function there, too as it is much less code.

Reply:
Funny story (not actually funny): I read the deprecation warning message ("use urlparse instead"), played with urlparse in a REPL, saw its `__repr__` which is basically

```
>>> urlparse('http://servername:443')
ParseResult(scheme='http', netloc='servername:443', path='', params='', query='', fragment='')
```

and said to myself "oh no, urlparse doesn't split hostname from port! it's totally unsuitable!".

Then I actually looked at your code in ZPublisher and saw that urlparse does in fact have `.hostname` and `.port` attributes, just doesn't show them in the `__repr__`.

(But then I decided that the regex based approach actually used by the stdlib's splitport() was less code than calling urlparse and fixing up API mismatches with IPv6 and types and everything.)